### PR TITLE
Modifications to support ludwig model

### PR DESCRIPTION
**Pull request by RanHomri**
Final changes for supporting Ludwig models

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "onnx2kerastl"
-version = "0.0.94"
+version = "0.0.95"
```
Comment (doronHarnoy) [inline pyproject.toml:3]: I think you would need to bump your version up to 96

File: onnx2kerastl/utils.py [hunk 2]
```diff
@@ -131,3 +132,38 @@ def check_torch_keras_error(model, k_model, input_np, epsilon=1e-5, change_order
...
+              If all tensors in the input list already have the same rank, the original list is returned.
+    """
+    ranks = [tensor.shape.rank for tensor in tensor_list]
+    max_rank = max(ranks)
+
+    if len(set(ranks)) == 1:
+        return tensor_list
+    elif len(set(ranks)) > 2:
+        raise ValueError(f"More than 2 different ranks detected, broadcasting is ambiguous.\n"
+                         f"Check the outputs of layers: \n" + '\n'.join([tensor.name for tensor in tensor_list]))
```
Comment (doronHarnoy) [inline onnx2kerastl/utils.py:158]: I think we can try to broadcast here multiple smaller ranks to the highest one no?

File: onnx2kerastl/utils.py [hunk 3]
```diff
@@ -21,8 +21,10 @@ def ensure_numpy_type(obj):
     """
     if is_numpy(obj):
         return obj
+    elif isinstance(obj, tf.Tensor):
+        return obj.numpy()
```
Comment (doronHarnoy) [inline onnx2kerastl/utils.py:25]: are you sure about it? i think we wanted to ensure with this function something is a hyper parameter, where does it blocks you?

File: onnx2kerastl/converter.py [hunk 4]
```diff
@@ -113,6 +113,8 @@ def onnx_to_keras(onnx_model, input_names, name_policy=None, verbose=True, chang
                 input_shape = [shape if shape != 0 else None for shape in input_shape]
                 batch_size = input_shape[0]
                 input_shape = input_shape[1:]
+                if len(input_shape) == 0:
+                    input_shape = 1
```
Comment (doronHarnoy) [inline onnx2kerastl/converter.py:117]: shouldn't the input shape alway be a list?

File: onnx2kerastl/converter.py [hunk 5]
```diff
@@ -197,16 +197,28 @@ def onnx_to_keras(onnx_model, input_names, name_policy=None, verbose=True, chang
...
+
+            # for case of weights sharing, map the shared weights to determine
+            # if a Gather layer is an embedding layer
+            if node_type == 'Identity' and node_input in weights:
+                embedding_weights_mapping[node_name] = node_input
+
+            # check conditions for embedding layer
+            is_in_weights = node_input in weights  # is this node input in weights
+            is_mapped_to_weights = embedding_weights_mapping.get(node_input, '') in weights  # is this node inputs weights are shared with other input
+            is_embedding = is_in_weights or is_mapped_to_weights  # if either is true this layer is a possible embedding layer
```
Comment (doronHarnoy) [inline onnx2kerastl/converter.py:213]: Cool!

**Review by doronHarnoy — APPROVED** [review]
Looks really good! I'm not sure if the fix to `ensure_numpy_type` making it ensuring numpy or tf tensor makes sense to me, I'd like to see when it is necessary
Regarding `match_tensors_rank` function, i feel like it would work and can be pushed in but I feel like we can handle general broadcasting in a broader way, maybe taking inspiration from how onnx is broadcasting

File: onnx2kerastl/utils.py (hunk 2 above)
Comment (RanHomri) [reply onnx2kerastl/utils.py:158]: I don't think so. If we have more than two different ranks than it becomes ambiguous which is the concatenation axis for each tensor, e.g., for tensor shapes (1, ), (1, 12), (2, 1, 4) and the concatenation axis is 1.
in this case how can we choose the reference tensor for the axis? should we broadcast the second tensor to (1, 12, 1) or to (1, 1, 12)?

**Pull request merged**